tests/helpers.py

```diff
@@ -0,0 +1,32 @@
+import tracemalloc
+
+class MemoryThreshold:
+    """
+    Context manager to check peak memory usage against an expected threshold.
+
+    example usage:
+    with MemoryThreshold(expected_usage):
+        # code that should not exceed expected
+    """
+
+    def __init__(self, expected_usage):
+        """
+        Parameters
+        ----------
+        expected_usage : int
+            Expected peak memory usage in bytes
+        """
+        self.expected_usage = expected_usage
+
+    def __enter__(self):
+        tracemalloc.start()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        _, peak = tracemalloc.get_traced_memory()
+        tracemalloc.stop()
+
+        assert peak <= self.expected_usage, (
```

Comment:
I'm on the fence about this being an `assert` vs a custom exception. Since it's only intended for testing `assert` seems fine but can you think of a time we might want to catch this in a test? If so, I'd say we make and use a custom exception.

Reply:
would the idea then be that the tests would error out instead of failing?

Reply:
The test would fail (not error) for a non-assert exception (assuming this class is used in a test not a fixture).

One place where we might want to use a custom exception would be in testing of this class.
```python
def test_memory_limit():
    with pytest.raises(MemoryThresholdExceeded):
        with MemoryThreshold(100):
            buff = b"\x00" * 1000
```
maybe the other place is if we use this as part of a fixture that also runs a pipeline (like the regtests in jwst). Given all the unknowns there it's probably not worth at this point trying to implement the needed features.

I can't think of a downside to making this a custom exception. I'm good either way.